src/Compilers/CSharp/Test/Symbol/Symbols/RequiredMembersTests.cs

```diff
@@ -3255,4 +3561,77 @@ public Derived() : this(true)
         );
     }
+
+    [Fact]
+    public void SetsRequiredMembersAppliedToRecordCopyConstructor_DeclaredInType()
+    {
+        var code = """
+            public record C
+            {
+                public required string Prop1 { get; set; }
+                public required int Field1;
+            }
+            """;
+
+        var comp = CreateCompilationWithRequiredMembers(code);
+        CompileAndVerify(comp, sourceSymbolValidator: validate, symbolValidator: validate);
+
+        void validate(ModuleSymbol module)
+        {
+            var c = module.GlobalNamespace.GetTypeMember("C");
+            var copyCtor = c.GetMembers(".ctor").Cast<MethodSymbol>().Single(m => m.ParameterCount == 1);
+
+            if (copyCtor is SynthesizedRecordCopyCtor)
+            {
+                Assert.Empty(copyCtor.GetAttributes());
+            }
+            else
+            {
+                AssertEx.Equal("System.Diagnostics.CodeAnalysis.SetsRequiredMembersAttribute..ctor()",
+                               copyCtor.GetAttributes().Single(a => a.AttributeClass!.IsWellKnownSetsRequiredMembersAttribute()).AttributeConstructor.ToTestDisplayString());
+            }
+        }
+    }
+
+    [Theory]
+    [CombinatorialData]
+    public void SetsRequiredMembersAppliedToRecordCopyConstructor_DeclaredInBase(bool useMetadataReference)
+    {
+        var @base = """
+            public record Base 
+            {
+                public required string Prop1 { get; set; }
+                public required int Field1;
+            }
+            """;
+
+        var code = """
+            public record Derived : Base;
+            """;
+
+        var comp = CreateCompilationWithRequiredMembers(new[] { @base, code });
+        CompileAndVerify(comp, sourceSymbolValidator: validate, symbolValidator: validate);
+
+        var baseComp = CreateCompilationWithRequiredMembers(@base);
+        CompileAndVerify(baseComp).VerifyDiagnostics();
+
+        comp = CreateCompilation(code, references: new[] { useMetadataReference ? baseComp.ToMetadataReference() : baseComp.EmitToImageReference() });
+        CompileAndVerify(comp, sourceSymbolValidator: validate, symbolValidator: validate);
+
+        void validate(ModuleSymbol module)
+        {
+            var c = module.GlobalNamespace.GetTypeMember("Derived");
+            var copyCtor = c.GetMembers(".ctor").Cast<MethodSymbol>().Single(m => m.ParameterCount == 1);
+
+            if (copyCtor is SynthesizedRecordCopyCtor)
+            {
+                Assert.Empty(copyCtor.GetAttributes());
+            }
+            else
+            {
+                AssertEx.Equal("System.Diagnostics.CodeAnalysis.SetsRequiredMembersAttribute..ctor()",
+                               copyCtor.GetAttributes().Single(a => a.AttributeClass!.IsWellKnownSetsRequiredMembersAttribute()).AttributeConstructor.ToTestDisplayString());
+            }
+        }
+    }
```

Comment:
Is there a test which exercises the new() constraint a la the following?

```cs
class C1
{
    public required string Prop { get; set; }

    [SetsRequiredMembers]
    public C1() { }
}

class C2
{
    public required string Prop { get; set; }

    public C2() { }
}

class Program
{
    static T Create<T>() where T : new() => new T();
    static void Main()
    {
        var c1 = Create<C1>(); // ok
        var c2 = Create<C2>(); // error
    }
}
```

Reply:
New constraints are the next PR :).